doc/release/upcoming_changes/26081.improvement.rst

```diff
@@ -0,0 +1,21 @@
+Shapes are now covariant
+------------------------
+When ``ndarray`` was made generic, more details and implementation were
+left to future development.  While it was recognized that shape was inherently
+a tuple (and thus covariant), development towards ``TypeVarTuple`` does not
+yet allow bounds or variance.  This change allows user code to type check in
+functions that attempt to specify the shape parameter:
+
+.. code::
+
+   from typing import NewType, Any
+   import numpy as np
+
+   Time = NewType("Time", int)
+   Series = NewType("Series", int)
+
+   arr: np.ndarray[tuple[Time, Series], Any] = np.arange(4.0).reshape((2, 2))
```

Comment:
I'd strongly prefer to add another type alias for shaped arrays to the `numpy.typing` namespace instead. Using plain `np.ndarray` as a type annotation (over `npt.NDArray`) is generally ill-advised due to the former's complexity, if it can be avoided that is. Something like a `ShapedArray` would be ideal. _i.e._ a type alias representing an array with a given shape *and* dtype; the docs should explicitly recommend `npt.NDArray` usage if the shape is unknown/arbitrary.

Note that prior to Python 3.11 people would have to make sure during runtime to either use `if TYPE_CHECKING:` blocks and/or `from __future__ import annotations` due to a lack of `TypeVarTuple` support. This caveat is perfectly acceptable though in my opinion.

```python
_ShapeType = TypeVarTuple("_ShapeType")
_ScalarType_co = TypeVar("_ScalarType_co", bound=generic, covariant=True)

if TYPE_CHECKING or sys.version_info >= (3, 11):
    ShapedArray = ndarray[tuple[*_ShapeType], dtype[_ScalarType_co]]
else:
    ShapedArray = ndarray[Any, dtype[_ScalarType_co]]
```

Reply:
@BvB93 I proposed exactly solution a couple of days ago https://github.com/numpy/numpy/issues/16544#issuecomment-2005169857. Great minds really do think alike, eh 😉 ?

Reply:
Happy to add an alias!  My only consideration is whether the alias should expect grouping dimensions, e.g.:
```python
FloatArray2D = npt.ShapedArray[tuple[int, int], np.float64]
```
I slightly agree with @vnmabus's [comment](1): "I think it [requiring dimensions to be grouped] would make it more consistent with functions and classes that need to take more than one shape".  That said, I'll defer to you if you want the flatter version of `ShapedArray`?

[1]: https://github.com/numpy/numpy/issues/16544#issuecomment-2006220209

Reply:
I don't think it should.

In my experience, it is very annoying to have to deal with long typing annotations. In this case, it is a differnce of 7 characters; close to 10% of the commonly 80 or 79 character width limit. To illustrate:

```python
npt.ShapedArray[tuple[int, int], np.float64]
npt.ShapedArray[int, int, np.float64]
```

If you still want to use "group" the shape types, then there's always  `np.ndarray[tuple[int, int], np.dtype[np.float64]]` to fall back to.

Reply:
But for the sake of readability, it might help to place the shapes *after* the dtype, e.g.

```python
type VectorF8[I: int] = npt.ShapedArray[np.float64, I]
type MatrixF8[I: int, J: int] = npt.ShapedArray[np.float64, I, J]
type CuboidF8[I: int, J: int, K: int] = npt.ShapedArray[np.float64, I, J, K]
```

I realize that this doesn't match the order of the `dtype` and `shape` arguments in e.g. `np.array`, but it might be more intuitive this way, since those used to `npt.NDArray` have it in their muscle memory to always start with the dtype. 

Furthermore, Python users are used to always having the `*args` after any positional arguments. So doing something different with the `*Shape` type parameter could be confusing.

Reply:
On a related note; what about naming it `Array` instead of `ShapedArray`?

Reply:
I think there's going to be a lot of opinions on variadic shape vs tuple shape and shape first vs dtype first.  Here's the pros/cons I've thought through so far:

#### variadic shape, dtype first:
* Follows example in PEP 646
* Similar to `foo(arg1, *args)`
  * People using `NDArray` are used to dtype first
* Shorter
#### tuple shape, dtype last:
* Similar to funcs like `np.ones(shape: tuple[int, ...], dtype: dtype)`
  * People who haven't used static typing are used to this order
* Same as `ndarray` typing
  * People who've already been static shape typing have to use this syntax
 
Polled my office and people unanimously supported the second.

**variadic shape, dtype last** seems like a decent compromise, and it's what I'll commit. I'm happy to sit on the PR and change it back if a better consensus is desired, but I've overthought myself to a point of analysis paralysis.  I've coded all these options up, hopped back and forth, and test-drove them in a few use cases (and they all seem to work fine)

> On a related note; how about naming it Array instead of ShapedArray?


Haha yeah good point.  ~thefacebook.com~ facebook.com :smile:

Reply:
Also a neat lesson learned here: 
```python
TupType = TypeVar("TupType", covariant=True, bound=tuple[int, ...])
TypeTup = TypeVarTuple("TypeTup")

class Foo[TupType]: ...
Bar = Foo[tuple[*TypeTup]]

def myfunc(a: Bar): ...
```
^ an effective way of enforcing type variance and bounds on a `TypeVarTuple`.  Combines (some of) the advantages of both.

Reply:
Actually, the unpacking restrictions on `tuple[int, ...]` mean that the third option isn't usable for certain functions, e.g. `def add_one_dimension(arr: Array[*Ts, Any]) -> Array[int, *Ts, Any]`

Reply:
It might help to return e.g. `Array[Any, *Ts, np.float64]`, instead of `Array[int, *Ts, np.float64]`. I'm not sure why, but `Any` results in covariant behaviour, whereas `int`  doesn't.

I agree that it would be better if you could `TypeVarTuple` as `covariant=True`, and `bound=int`, but unfortunately that's not implemented at the moment (for no valid reason whatsoever IMHO).

Reply:
> variadic shape, dtype last seems like a decent compromise, and it's what I'll commit. I'm happy to sit on the PR and change it back if a better consensus is desired, but I've overthought myself to a point of analysis paralysis. I've coded all these options up, hopped back and forth, and test-drove them in a few use cases (and they all seem to work fine)

I'm personally in favor of this approach, yes. The main goal of this type alias is to provide a _convenient_ alias for the end-user that's less verbose than the full syntax, providing syntactic sugar for turning something like `np.ndarray[tuple[int, int], np.dtype[np.float64]` into a more manageable `npt.Array[int, int, np.float64]`. 

Are there edge cases wherein this alias cannot be used? As you have noted previously: yes, yes there are. Is this a problem? No, not as long as it's usable in the majority of cases; in all others one can always fall back to plain` np.ndarray` usage (something which holds as well for `npt.NDArray`).